Altimeter/Src/altimeter.cpp

```diff
@@ -0,0 +1,166 @@
+}
+
+
+
+uint32_t AltDevice::uncompensatedPressureTemperature(SPI_HandleTypeDef *spi_handle, GPIO_TypeDef *cs_port, uint16_t cs_pin, uint8_t conversion_command){
+
+	uint8_t digital_pres_temp[1] = {conversion_command};
+	uint8_t receive_buffer[1];
+
+	HAL_GPIO_WritePin(cs_port, cs_pin, GPIO_PIN_SET);
+	HAL_GPIO_WritePin(cs_port, cs_pin, GPIO_PIN_RESET);
+
+	/* send the digital pressure or temperature with the command */
+	HAL_SPI_TransmitReceive(spi_handle, digital_pres_temp, receive_buffer, 1, TIMEOUT);
+	HAL_GPIO_WritePin(cs_port, cs_pin, GPIO_PIN_SET);
+
+	/* at least 8.22 ms delay required for ADC conversion.*/
+	osDelay(9);
+
+	uint8_t tx_buffer[4] = {ADC_READ, 0xFF, 0xFF, 0xFF};
+	uint8_t rx_buffer[4];
+
+	HAL_GPIO_WritePin(cs_port, cs_pin, GPIO_PIN_RESET);
+
+	/*Start adc conversion to get value */
+	HAL_SPI_TransmitReceive(spi_handle, tx_buffer, rx_buffer, 4, TIMEOUT);
+
+	HAL_GPIO_WritePin(cs_port, cs_pin, GPIO_PIN_SET);
+
+	//get 24-bit uncompensated pressure or temperature value value.
+	uint32_t uncompensated_value = (rx_buffer[1] << 16 | rx_buffer[2] << 8 | rx_buffer[3]);
+
+	return uncompensated_value;
+}
+
+
+void AltDevice::calculateTempPres(SPI_HandleTypeDef *spi_handle, GPIO_TypeDef *cs_port, uint16_t cs_pin){
+
+
+	uint32_t digital_temperature = uncompensatedPressureTemperature(spi_handle, cs_port, cs_pin, CONVERT_D2_OSR_256);
+	uint32_t digital_pressure = uncompensatedPressureTemperature(spi_handle, cs_port, cs_pin, CONVERT_D1_OSR_256);
+
+	float dt = digital_temperature - (float)coeffs_.t_ref * 256.0f;
+	float temp = 2000 + dt * (float) coeffs_.temp_sens / pow(2, 23);
+
+
+	float off = coeffs_.off * 65536.0f + (coeffs_.tco * dt) / 128.0f;
+	float sens = coeffs_.sens * 32768.0f  + (coeffs_.tcs * dt) / 256.0f;
+
+	/*Second order temperature compensation */
+	if(temp < 2000){
+		float t2 = (dt*dt) / 2147483648.0f;
+		float off2 = 5.0f * pow((temp - 2000), 2) / 2.0f;
+		float sens2 = 5.0f * pow((temp - 2000), 2) / 4.0f;
+
+
+		if(temp < 1500){
+			off2 = off2 + 7 * pow((temp + 1500), 2);
+			sens2 = sens2 + 11 * pow((temp + 1500), 2) / 2;
+		}
+
+		temp = temp - t2;
+		off = off - off2;
+		sens = sens - sens2;
+
+	}
+
+	float pressure = (digital_pressure * (sens / 2097152.0f) - off) / 32768.0f;
+	pressure /= 100.0f;
+	temp /= 100.0f;
+
+	pres_ = pressure;
+	temp_ = temp;
+}
+
+
+float AltDevice::getPressure(){
+	return pres_;
+}
+
+float AltDevice::getTemperature(){
+	return temp_;
+}
+
+float AltDevice::getAltAboveSeaLvl(){
+
+	float reference_temp = 288.15f; /* Ref temperature in Kelvins */
+	float temp_lapse_rate = 0.0065f;
+	float exp_gmrl = 5.2558;
+
+	float reference_pressure = 101325.0f; /* in Pa */
+	float current_pressure = getPressure() * 100.0f;
+	float exponent = (log(current_pressure) - log(reference_pressure)) / exp_gmrl;
+
+	float height = reference_temp/temp_lapse_rate * (1 - pow(M_E, exponent));
+
+	return height;
+}
+
+
+float AltDevice::getAltitude(){
+	height_ = getAltAboveSeaLvl();
```

Comment:
is height_ a private var? Why are we modifying a private var here but not in the above function?